src/lib/homestore/heap_chunk_selector.cpp

```diff
@@ -0,0 +1,131 @@
+#include "heap_chunk_selector.h"
+
+#include <execution>
+#include <algorithm>
+#include <utility>
+
+#include <sisl/logging/logging.h>
+
+namespace homeobject {
+// https://github.com/eBay/HomeObject/pull/30#discussion_r1331112743
+// we make the following assumptions
+// 1 homestore will initialize HeapChunkSelector by adding all the chunks single threaded
+// 2 we do not need dynamic chunk requirements
+
+// it means after the single thread initialization,
+// 1 the key collection of m_per_dev_heap will never change.
+// 2 the key collection of m_chunks will never change
+void HeapChunkSelector::add_chunk(csharedChunk& chunk) {
+    VChunk vchunk(chunk);
+    auto pdevID = vchunk.get_pdev_id();
+    // add this find here, since we don`t want to call make_shared in try_emplace every time.
+    auto it = m_per_dev_heap.find(pdevID);
+    if (it == m_per_dev_heap.end()) it = m_per_dev_heap.emplace(pdevID, std::make_shared< PerDevHeap >()).first;
+    auto& avalableBlkCounter = it->second->available_blk_count;
+    avalableBlkCounter.fetch_add(vchunk.available_blks());
+
+    m_chunks.emplace(vchunk.get_chunk_id(), chunk);
+
+    auto& heapLock = it->second->mtx;
+    auto& heap = it->second->m_heap;
+    std::lock_guard< std::mutex > l(heapLock);
+    heap.emplace(chunk);
+}
+
+// select_chunk will only be called in homestore when creating a shard.
+csharedChunk HeapChunkSelector::select_chunk(homestore::blk_count_t count, const homestore::blk_alloc_hints& hint) {
```

Comment:
`count` is not used anywhere in this function. If this count is meant for seleted_chunk should at least have more than this number of blks, you probably need to compare it before return, or is it for any other puspose?

Reply:
this is just a function from homestore, and i just derived from it. I have marked it as `[[maybe_unused]]` in the header file. if we want to change this , we need to change the header file in homestore.

Reply:
The count hint is for selector to pick up a chunk that "best-effort" fulfill the `count` blks requirement.  
HeapChunkSelector by nature implemented this as it always pick the largest chunk so if the chunk picked doesnt have `count` blks then no other chunk can have `count` blks.

However other selector doesnt behave like this, for example the round-robin selector.  It should best check size before returning.

Reply:
it will be weird for the consumer to provide a `count` when calling heap to select a chunk.
We can make change to API signature from both sides later.